enhancements/update/eus-upgrades-mvp.md

```diff
@@ -0,0 +1,296 @@
+analyze upgrade timings using simulated workloads at a defined capacity and
+load. This analysis will be used to identify optimizations and set expectations
+on a per minor version basis as to the duration of the upgrade.
+
+Due to the variability in workload deployment these measurements will be
+*exclusive of Worker MCP rollout*. The simulated workload is intended only to
+ensure that our measurements more closely mimic real world usage, not upgrades
+of clusters with zero workload.
+
+#### As an admin, upgrades of CVO managed components are as fast as safe
+
+In 4.6 and 4.7 we identified a number of DaemonSets which were not critical to
+workload availability which could rollout in a more parallel manner. There are a
+few remaining DaemonSets which could be made more parallel after confirming a
+canary deployment passes such as multus, sdn, ovs.
+
+Once these new deployment patterns have been vetted in 4.8 or 4.9 these changes
+should be backported through to 4.6.
+
+#### As an admin, my workloads within a failure domain are rescheduled a minimum number of times
+
+Currently the MCO cycles nodes in a deterministic but unintentional manner. The
+workload on that host which can be relocated scatters to other hosts according
+to scheduling rules. This has been measured and shown that approximately 70% of
+all reschedulable pods are rescheduled more than necessary if they were
+rescheduled in an optimal manner. Optimal rescheduling should be closer to (1 /
+node_count_per_failure_domain) percentage of pods being rescheduled more than
+once during a rolling restart of a given failure domain.
+
+We should make the MCO topology aware so that it upgrades all of the host in one
+failure domain because anti-affinity scheduling is used in most scenarios. We
+should provide some mechanism to rank upgraded nodes as slightly higher priority
+in scheduler configuration. Perhaps a timestamp of configuration generation that
+the scheduler then biases toward nodes with the greatest timestamps as a last in
+chain priority.
+
+We should consider enabling the rescheduler as well to ensure that when the
+upgrade completes we rebalance so that the last node ends up with proportionate
+workload.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The main caveat of what's outlined here is that it does not reduce reboots
+which, for many reasons, is highly desired by some admins. The rationale for not
+addressing that in this enhancement is that this enhancement targets
+improvements which benefit all upgrades.
+
+The epic / user story to backport API removal notifications to 4.6 assumes that
+we will require anyone upgrading along an EUS specific upgrade path to update to
+a 4.6 version which is specific to EUS. This is necessary because only those who
+follow the EUS upgrade path have 4.10 as their next logical target version. We
+should not alert non EUS clusters of removals in 4.10 until they've upgraded to
```

Comment:
Maybe this is the intention driven experience I was asking about above.  It isn't that they've picked 4.10 (they still have to pick 4.7 as the next target) it's that they've upgraded to this 4.6 EUS version that has all the notification items backported to it?

In that case would there be an EUS 4.7, 4.8, and 4.9 or would the backported notifications show for people not on the EUS path too? (or the alternative above that only shows {next releases notifications} with expansion to {next EUS notifications}

Reply:
The backporting to 4.7, 4.8, and 4.9 is a good point to raise. I hadn't thought about that, I would assume we leave alerting rules in those versions whatever is appropriate for non-EUS upgrades. We know that 4.6 EUS upgrades will be upgrading through those versions but if we raise the red flags in 4.6 EUS versions and they start the upgrade anyway I feel like they were given enough warning.

I'm also envisioning an escalation of alerts based on when the API is removed. APIs removed in N+1 are warning, N+2 or greater are info level only. When running an EUS specific version N+1 should be equivalent to the alert set of your eventual target minus one, ie: 4.9 in the case that 4.6 and 4.10 are the two EUS versions.

Reply:
I think now that we know we'll only use Info level alerts for all removals between 4.6 and 4.10 and then Upgradeable=False for removals between 4.N and 4.N+1 we no longer require this intent signal. It's safe to tell all 4.6 customers about removals between 4.6 and 4.10 and it's safe/required to inhibit upgrades from 4.6 to 4.7 when APIs removed in 4.7 are in use.